innopoints/views/application.py

```diff
@@ -77,3 +92,85 @@ def take_back_application(project_id, activity_id):
         abort(400, {'message': 'Data integrity violated.'})
 
     return NO_PAYLOAD
+
+
+# ----- VolunteeringReport -----
+
+@api.route('/projects/<int:project_id>/activities/<int:activity_id>'
+           '/applications/<int:application_id>/report_info')
+@login_required
+def get_report_info(project_id, activity_id, application_id):
+    """Get the reports from the moderators of the project and an average rating."""
+    application = Application.query.get_or_404(application_id)
+    activity = Activity.query.get_or_404(activity_id)
+    project = Project.query.get_or_404(project_id)
+
+    if activity.project != project or application.activity_id != activity.id:
+        abort(400, {'message': 'The specified project, activity and application are unrelated.'})
+
+    if current_user not in project.moderators and not current_user.is_admin:
+        abort(401)
+
+    avg_rating = db.session.query(
+        db.func.round(db.func.avg(VolunteeringReport.rating))
+    ).join(Application).join(Activity).join(
+        project_moderation,
+        VolunteeringReport.reporter_email == project_moderation.c.account_email
+    ).filter(
+        Application.applicant_email == application.applicant_email,
+        project_moderation.c.project_id == project_id,
+    ).scalar() or 0
+
+    reports = VolunteeringReport.query.join(Application).join(Activity).join(
+        project_moderation,
+        VolunteeringReport.reporter_email == project_moderation.c.account_email
+    ).filter(
+        Application.applicant_email == application.applicant_email,
+        project_moderation.c.project_id == project_id,
+    ).all()
+
+    out_schema = VolunteeringReportSchema(only=('content', 'rating', 'time'), many=True)
+    return jsonify(average_rating=int(avg_rating), reports=out_schema.dump(reports))
+
+
+@api.route('/projects/<int:project_id>/activities/<int:activity_id>'
+           '/applications/<int:application_id>/report', methods=['POST'])
+@login_required
+def create_report(project_id, activity_id, application_id):
+    """Create a volunteering report on an application."""
+    if not request.is_json:
+        abort(400, {'message': 'The request should be in JSON.'})
+
+    application = Application.query.get_or_404(application_id)
+    activity = Activity.query.get_or_404(activity_id)
+    project = Project.query.get_or_404(project_id)
+
+    if activity.project != project or application.activity_id != activity.id:
+        abort(400, {'message': 'The specified project, activity and application are unrelated.'})
+
+    if current_user not in project.moderators and not current_user.is_admin:
+        abort(401)
+
+    if not (project.lifetime_stage == LifetimeStage.ongoing
+            and project.review_status is not None):
+        abort(400, {'message': 'The project must be in the finalizing stage.'})
+
+    in_schema = VolunteeringReportSchema(exclude=('time',))
```

Comment:
Wouldn't setting 'time' to be dump_only allow us to not write it in the `exclude` every time the schema is used for loading only?

Reply:
Either work, I just always did it like this

Reply:
I see, but I think one looks cleaner than the other. At least for future code)